pkg/controller/deployment/util/deployment_util.go

```diff
@@ -51,6 +51,10 @@ const (
 	// is deployment.spec.replicas + maxSurge. Used by the underlying replica sets to estimate their
 	// proportions in case the deployment has surge replicas.
 	MaxReplicasAnnotation = "deployment.kubernetes.io/max-replicas"
+	// FailedProgressAnnotation on a replica set denotes that it has failed to progress during its deployment.
+	FailedProgressAnnotation = "deployment.kubernetes.io/failed-progressing"
+	// CompleteProgressAnnotation on a replica set denotes that it has completed successfully its deployment.
+	CompleteProgressAnnotation = "deployment.kubernetes.io/complete-progressing"
```

Comment:
Hm, actually we can't maintain all the history for a deployment in its conditions (imagine a deployment with hundreds of replica sets). We need these:)